ui/apps/platform/src/Containers/Integrations/utils/integrationUtils.ts

```diff
@@ -74,3 +76,70 @@ const getTimes = () => {
 };
 
 export const timesOfDay = getTimes();
+
+// Utilities for image integrations which can have either or both of two categories.
+
+// Categories alternatives correspond to mutually exclusive toggle group items.
+type CategoriesAlternatives<
+    Category0 extends ImageIntegrationCategory,
+    Category1 extends ImageIntegrationCategory
+> = [
+    [[category0: Category0]],
+    [[category1: Category1]],
+    // The alternative for both categories includes both orders.
+    [[category0: Category0, category1: Category1], [category1: Category1, category0: Category0]]
+];
+
+// Compiler verifies that first argument of matchCategoriesAlternative method is a category alternative.
+type CategoriesAlternative<
+    Category0 extends ImageIntegrationCategory,
+    Category1 extends ImageIntegrationCategory
+> = CategoriesAlternatives<Category0, Category1>[number];
+
+function getCategoriesUtils<
+    Category0 extends ImageIntegrationCategory,
+    Category1 extends ImageIntegrationCategory
+>([category0, category1]: [Category0, Category1], [text0, text1]: [string, string]) {
+    const categoriesAlternatives: CategoriesAlternatives<Category0, Category1> = [
+        [[category0]],
+        [[category1]],
+        [
+            [category0, category1],
+            [category1, category0],
+        ],
+    ];
+
+    // For robust behavior, do not assume that categories from response are limited to Category0 and Category1.
+    /* eslint-disable no-nested-ternary */
+    return {
+        categoriesAlternatives,
+
+        getCategoriesText: (categories: ImageIntegrationCategory[]) =>
```

Comment:
I see the comment about robust behavior above, but is it possible to replace `ImageIntegrationCategory[]` here and below with `(Category0 | Category1)[]`? From what I can tell all usages of this are from client side code which can guarantee the the array will only contain the items passed to `getCategoriesUtils`.

Disregard if I missed another place where this function is called.

Reply:
Good eyes. This possible improvement will simmer in my slow cooker until step 2 below.

1. The first scenario is editing an existing integration, in which the starting value is from response.
2. The next pull request (which you cannot have missed) will add a **Type** column to the 3 integration tables. At the moment, TypeScript would report an error because of `categories: ImageIntegrationCategory[]` in `BaseImageIntegration` type. If I can override the property type for each of the 3 derived types, then I will change the argument type of the function, and keep the fallback case, just in case. I verified that TypeScript does not complain about the redundancy.